api/api/middlewares.py

```diff
@@ -153,83 +174,99 @@ async def check_rate_limit(
-    'await handler(request) it means the output will be a 200 response and no fields needs to be removed.
-    """
-
-    def remove_unwanted_fields(fields_to_remove=None):
-        fields_to_remove = fields_to_remove or ['status', 'type']
-        for field in fields_to_remove:
-            if field in problem.body:
-                del problem.body[field]
-        if problem.body.get('detail') == '':
-            del problem.body['detail']
-        if 'code' in problem.body:
-            problem.body['error'] = problem.body.pop('code')
-
-    problem = None
-
-    try:
-        return await handler(request)
-
-    except ProblemException as ex:
-        problem = connexion_problem(status=ex.__dict__['status'],
-                                    title=ex.__dict__['title'] if ex.__dict__.get('title') else 'Bad Request',
-                                    type=ex.__dict__.get('type', 'about:blank'),
-                                    detail=_cleanup_detail_field(ex.__dict__['detail'])
-                                    if 'detail' in ex.__dict__ else '',
-                                    ext=ex.__dict__.get('ext'))
-    except HTTPException as ex:
-        problem = connexion_problem(ex.status,
-                                    ex.reason if ex.reason else '',
-                                    type=ex.reason if ex.reason else '',
-                                    detail=ex.text if ex.text else '')
-    except (OAuthProblem, Unauthorized) as auth_exception:
-        if request.path in {'/security/user/authenticate', '/security/user/authenticate/run_as'} and \
-                request.method in {'GET', 'POST'}:
-            await prevent_bruteforce_attack(request=request, attempts=api_conf['access']['max_login_attempts'])
-            problem = connexion_problem(401, "Unauthorized", type="about:blank", detail="Invalid credentials")
+        if error_code:
+            raise MaxRequestsException(code=error_code)
         else:
-            if isinstance(auth_exception, OAuthProblem):
-                problem = connexion_problem(401, "Unauthorized", type="about:blank",
-                                            detail="No authorization token provided")
-            else:
-                problem = connexion_problem(401, "Unauthorized", type="about:blank", detail="Invalid token")
-    finally:
-        problem and remove_unwanted_fields()
-
-    return problem
+            return await call_next(request)
+
+
+class CheckBlockedIP(BaseHTTPMiddleware):
+    """Rate Limits Middleware."""
+
+    async def dispatch(self, request: Request, call_next: RequestResponseEndpoint) -> Response:
+        """"Update and check if the client IP is locked."""
+        if request.url.path in {LOGIN_ENDPOINT, RUN_AS_LOGIN_ENDPOINT} \
+           and request.method in {'GET', 'POST'}:
+            check_blocked_ip(request)
+        return await call_next(request)
```

Comment:
With the configuration below:
```YAML
# Access parameters
access:
 max_login_attempts: 5
 block_time: 20
 max_request_per_minute: 10
```

I'm getting some errors when the IP should be blocked after failing 5 auth attempts:
```
2024/02/19 12:10:33 INFO: Starting API in foreground
2024/02/19 12:10:33 INFO: Checking RBAC database integrity...
2024/02/19 12:10:33 INFO: /var/ossec/api/configuration/security/rbac.db file was detected
2024/02/19 12:10:33 INFO: RBAC database integrity check finished successfully
2024/02/19 12:10:38 INFO: Listening on 0.0.0.0:55000.
2024/02/19 12:10:41 INFO: wazuh 192.168.48.1 "POST /security/user/authenticate" with parameters {} and body {} done in 0.221s: 401
2024/02/19 12:10:41 INFO: wazuh 192.168.48.1 "POST /security/user/authenticate" with parameters {} and body {} done in 0.168s: 401
2024/02/19 12:10:42 INFO: wazuh 192.168.48.1 "POST /security/user/authenticate" with parameters {} and body {} done in 0.164s: 401
2024/02/19 12:10:43 INFO: wazuh 192.168.48.1 "POST /security/user/authenticate" with parameters {} and body {} done in 0.166s: 401
2024/02/19 12:10:44 INFO: wazuh 192.168.48.1 "POST /security/user/authenticate" with parameters {} and body {} done in 0.172s: 401
2024/02/19 12:10:44 ERROR: Exception in ASGI application
  + Exception Group Traceback (most recent call last):
  |   File "/var/ossec/framework/python/lib/python3.10/site-packages/starlette/_utils.py", line 82, in collapse_excgroups
  |     yield
  |   File "/var/ossec/framework/python/lib/python3.10/site-packages/starlette/middleware/base.py", line 190, in __call__
  |     async with anyio.create_task_group() as task_group:
  |   File "/var/ossec/framework/python/lib/python3.10/site-packages/anyio/_backends/_asyncio.py", line 678, in __aexit__
  |     raise BaseExceptionGroup(
  | exceptiongroup.ExceptionGroup: unhandled errors in a TaskGroup (1 sub-exception)
  +-+---------------- 1 ----------------
    | Traceback (most recent call last):
    |   File "/var/ossec/framework/python/lib/python3.10/site-packages/starlette/middleware/errors.py", line 164, in __call__
    |     await self.app(scope, receive, _send)
    |   File "/var/ossec/framework/python/lib/python3.10/site-packages/starlette/middleware/base.py", line 189, in __call__
    |     with collapse_excgroups():
    |   File "/var/ossec/framework/python/lib/python3.10/contextlib.py", line 153, in __exit__
    |     self.gen.throw(typ, value, traceback)
    |   File "/var/ossec/framework/python/lib/python3.10/site-packages/starlette/_utils.py", line 88, in collapse_excgroups
    |     raise exc
    |   File "/var/ossec/framework/python/lib/python3.10/site-packages/starlette/middleware/base.py", line 191, in __call__
    |     response = await self.dispatch_func(request, call_next)
    |   File "/var/ossec/framework/python/lib/python3.10/site-packages/api/middlewares.py", line 237, in dispatch
    |     response = await call_next(request)
    |   File "/var/ossec/framework/python/lib/python3.10/site-packages/starlette/middleware/base.py", line 165, in call_next
    |     raise app_exc
    |   File "/var/ossec/framework/python/lib/python3.10/site-packages/starlette/middleware/base.py", line 151, in coro
    |     await self.app(scope, receive_or_disconnect, send_no_error)
    |   File "/var/ossec/framework/python/lib/python3.10/site-packages/connexion/middleware/exceptions.py", line 101, in __call__
    |     await super().__call__(scope, receive, send)
    |   File "/var/ossec/framework/python/lib/python3.10/site-packages/starlette/middleware/exceptions.py", line 62, in __call__
    |     await wrap_app_handling_exceptions(self.app, conn)(scope, receive, send)
    |   File "/var/ossec/framework/python/lib/python3.10/site-packages/starlette/_exception_handler.py", line 63, in wrapped_app
    |     response = await handler(conn, exc)
    |   File "/var/ossec/framework/python/lib/python3.10/site-packages/connexion/middleware/exceptions.py", line 37, in wrapper
    |     response = await handler(request, exc)  # type: ignore
    |   File "/var/ossec/framework/python/lib/python3.10/site-packages/api/error_handler.py", line 196, in problem_error_handler
    |     problem.update(exc.ext if exc.ext else {})
    | TypeError: 'ConnexionRequest' object is not subscriptable
    +------------------------------------

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/var/ossec/framework/python/lib/python3.10/site-packages/starlette/middleware/errors.py", line 164, in __call__
    await self.app(scope, receive, _send)
  File "/var/ossec/framework/python/lib/python3.10/site-packages/starlette/middleware/base.py", line 189, in __call__
    with collapse_excgroups():
  File "/var/ossec/framework/python/lib/python3.10/contextlib.py", line 153, in __exit__
    self.gen.throw(typ, value, traceback)
  File "/var/ossec/framework/python/lib/python3.10/site-packages/starlette/_utils.py", line 88, in collapse_excgroups
    raise exc
  File "/var/ossec/framework/python/lib/python3.10/site-packages/starlette/middleware/base.py", line 191, in __call__
    response = await self.dispatch_func(request, call_next)
  File "/var/ossec/framework/python/lib/python3.10/site-packages/api/middlewares.py", line 237, in dispatch
    response = await call_next(request)
  File "/var/ossec/framework/python/lib/python3.10/site-packages/starlette/middleware/base.py", line 165, in call_next
    raise app_exc
  File "/var/ossec/framework/python/lib/python3.10/site-packages/starlette/middleware/base.py", line 151, in coro
    await self.app(scope, receive_or_disconnect, send_no_error)
  File "/var/ossec/framework/python/lib/python3.10/site-packages/connexion/middleware/exceptions.py", line 101, in __call__
    await super().__call__(scope, receive, send)
  File "/var/ossec/framework/python/lib/python3.10/site-packages/starlette/middleware/exceptions.py", line 62, in __call__
    await wrap_app_handling_exceptions(self.app, conn)(scope, receive, send)
  File "/var/ossec/framework/python/lib/python3.10/site-packages/starlette/_exception_handler.py", line 63, in wrapped_app
    response = await handler(conn, exc)
  File "/var/ossec/framework/python/lib/python3.10/site-packages/connexion/middleware/exceptions.py", line 37, in wrapper
    response = await handler(request, exc)  # type: ignore
  File "/var/ossec/framework/python/lib/python3.10/site-packages/api/error_handler.py", line 196, in problem_error_handler
    problem.update(exc.ext if exc.ext else {})
TypeError: 'ConnexionRequest' object is not subscriptable

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/var/ossec/framework/python/lib/python3.10/site-packages/uvicorn/protocols/http/h11_impl.py", line 408, in run_asgi
    result = await app(  # type: ignore[func-returns-value]
  File "/var/ossec/framework/python/lib/python3.10/site-packages/uvicorn/middleware/proxy_headers.py", line 84, in __call__
    return await self.app(scope, receive, send)
  File "/var/ossec/framework/python/lib/python3.10/site-packages/connexion/apps/abstract.py", line 284, in __call__
    return await self.middleware(scope, receive, send)
  File "/var/ossec/framework/python/lib/python3.10/site-packages/connexion/middleware/main.py", line 501, in __call__
    await self.app(scope, receive, send)
  File "/var/ossec/framework/python/lib/python3.10/site-packages/starlette/middleware/errors.py", line 181, in __call__
    await response(scope, receive, send)
TypeError: 'coroutine' object is not callable
```

Reply:
I fixed the problem in a new commit. I reproduced the failure by running the integration `test_api/test_config/test_bruteforce_blocking_system`. The test now pass OK.

```
root@b843e90c07fb:/wazuh/tests/integration/test_api# pytest -vv test_config/test_bruteforce_blocking_system/
============================================================================================= test session starts ==============================================================================================
platform linux -- Python 3.10.12, pytest-7.1.2, pluggy-1.4.0 -- /usr/bin/python3
cachedir: .pytest_cache
metadata: {'Python': '3.10.12', 'Platform': 'Linux-5.15.133.1-microsoft-standard-WSL2-x86_64-with-glibc2.35', 'Packages': {'pytest': '7.1.2', 'pluggy': '1.4.0'}, 'Plugins': {'html': '3.1.1', 'metadata': '3.1.1'}}
rootdir: /wazuh/tests/integration, configfile: pytest.ini
plugins: html-3.1.1, metadata-3.1.1
collected 1 item

test_config/test_bruteforce_blocking_system/test_bruteforce_blocking_system.py::test_bruteforce_blocking_system[BRUTE_FORCE_BLOCKING_TIME] PASSED                                                        [100%]

============================================================================================== 1 passed in 28.20s ==============================================================================================
```